nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullHandler.java

```diff
@@ -0,0 +1,211 @@
+    Set<String> overriddenFieldNames = getAnnotationValueArray(overriddenMethod, annotName, false);
+    if (overriddenFieldNames == null) {
+      return;
+    }
+    if (overridingFieldNames == null) {
+      overridingFieldNames = Collections.emptySet();
+    }
+    if (overridingFieldNames.containsAll(overriddenFieldNames)) {
+      return;
+    }
+    overriddenFieldNames.removeAll(overridingFieldNames);
+
+    StringBuilder errorMessage = new StringBuilder();
+    errorMessage
+        .append(
+            "postcondition inheritance is violated, this method must guarantee that all fields written in the @EnsuresNonNull annotation of overridden method ")
+        .append(ASTHelpers.enclosingClass(overriddenMethod).getSimpleName())
+        .append(".")
+        .append(overriddenMethod.getSimpleName())
+        .append(" are @NonNull at exit point as well. Fields [");
+    Iterator<String> iterator = overriddenFieldNames.iterator();
+    while (iterator.hasNext()) {
+      errorMessage.append(iterator.next());
+      if (iterator.hasNext()) {
+        errorMessage.append(", ");
+      }
+    }
+    errorMessage.append(
+        "] must explicitly appear as parameters at this method @EnsuresNonNull annotation");
+    state.reportMatch(
+        analysis
+            .getErrorBuilder()
+            .createErrorDescription(
+                new ErrorMessage(
+                    ErrorMessage.MessageTypes.WRONG_OVERRIDE_POSTCONDITION,
+                    errorMessage.toString()),
+                tree,
+                buildDescriptionFromChecker(tree, analysis),
+                state));
+  }
+
+  /**
+   * On every method annotated with {@link com.uber.nullaway.qual.EnsuresNonNull}, this method,
+   * injects the {@code Nonnull} value for the class fields given in the {@code @EnsuresNonNull}
+   * parameter to the dataflow analysis.
+   */
+  @Override
+  public NullnessHint onDataflowVisitMethodInvocation(
+      MethodInvocationNode node,
+      Types types,
+      Context context,
+      AccessPathNullnessPropagation.SubNodeValues inputs,
+      AccessPathNullnessPropagation.Updates thenUpdates,
+      AccessPathNullnessPropagation.Updates elseUpdates,
+      AccessPathNullnessPropagation.Updates bothUpdates) {
+    if (node.getTree() == null) {
+      // A synthetic node might be inserted by the Checker Framework during CFG construction, it is
+      // safer to do a null check here.
+      return super.onDataflowVisitMethodInvocation(
+          node, types, context, inputs, thenUpdates, elseUpdates, bothUpdates);
+    }
+    Symbol.MethodSymbol methodSymbol = ASTHelpers.getSymbol(node.getTree());
+    Preconditions.checkNotNull(methodSymbol);
+    Set<String> fieldNames = getAnnotationValueArray(methodSymbol, annotName, false);
+    if (fieldNames == null) {
+      return super.onDataflowVisitMethodInvocation(
+          node, types, context, inputs, thenUpdates, elseUpdates, bothUpdates);
+    }
```

Comment:
It might be a bit cleaner to invert this condition:

```
if (fieldNames != null) {
 [All the extra handling]
}
return super.onDataflowVisitMethodInvocation(
          node, types, context, inputs, thenUpdates, elseUpdates, bothUpdates);
```

But it's probably fine either way :)

Reply:
https://github.com/uber/NullAway/pull/423/commits/bf8d1ad295578d1f54608f7d75d2fdf20a06e25b